File: php-core/src/TokenCache.php
<?php

declare(strict_types=1);

namespace Telebirr\Sdk\Core;

/**
 * In-memory token cache with configurable time-to-live (TTL).
 *
 * Stores a fabric token in memory and automatically invalidates it
 * after the configured TTL expires.
 */
class TokenCache
{
    /** @var string|null The cached fabric token, or null if not cached */
    private ?string $token = null;

    /** @var int Unix timestamp (seconds) when the cached token expires */
    private int $expiresAt = 0;

    /** @var int Time-to-live in milliseconds */
    private readonly int $ttlMs;

    /**
     * Create a new TokenCache with the specified TTL.
     *
     * @param int $ttlMs Time-to-live in milliseconds (default: 3000000, i.e. 50 minutes)
     */
    public function __construct(int $ttlMs = 3000000)
    {
        $this->ttlMs = $ttlMs;
    }

    /**
     * Retrieve the cached token if it is still valid.
     *
     * Returns the token if cached and not expired; otherwise returns null
     * and clears the internal cache state.
     *
     * @return string|null The cached fabric token, or null if expired or not set
     */
    public function get(): ?string
    {
        if ($this->token === null || time() >= $this->expiresAt) {
            $this->token = null;
            return null;
        }
        return $this->token;
    }

    /**
     * Store a token in the cache with the configured TTL.
     *
     * @param string $token The fabric token to cache
     *
     * @return void
     */
    public function set(string $token): void
    {
        $this->token = $token;
        $this->expiresAt = time() + ($this->ttlMs / 1000);
    }

    /**
     * Clear the cached token and reset the expiry.
     *
     * @return void
     */
    public function clear(): void
    {
        $this->token = null;
        $this->expiresAt = 0;
    }
}

File: dart_core/lib/src/token_cache.dart
/// An in-memory token cache with configurable time-to-live (TTL).
///
/// Stores the fabric token obtained from the Telebirr token endpoint and
/// automatically invalidates it after the configured TTL expires. This
/// prevents unnecessary token refresh requests.
///
/// The default TTL is 50 minutes, which is slightly less than the typical
/// token lifetime to ensure tokens are refreshed before they expire.
///
/// Example:
/// ```dart
/// final cache = TokenCache(ttl: Duration(minutes: 45));
/// cache.set('my-fabric-token');
/// final token = cache.get(); // Returns the token if not expired
/// ```
class TokenCache {
  /// The cached fabric token, or `null` if no token is cached.
  String? _token;

  /// The expiration timestamp for the cached token, or `null` if empty.
  DateTime? _expiresAt;

  /// The time-to-live duration for cached tokens.
  ///
  /// After this duration from [set], the token is considered expired
  /// and [get] returns `null`.
  final Duration ttl;

  /// Creates a [TokenCache] with an optional custom [ttl].
  ///
  /// Defaults to 50 minutes if not specified.
  TokenCache({Duration? ttl}) : ttl = ttl ?? const Duration(minutes: 50);

  /// Returns the cached token if it exists and has not expired.
  ///
  /// Returns `null` if:
  /// - No token has been set.
  /// - The token has expired (TTL elapsed since [set] was called).
  ///
  /// When the token is expired, it is automatically cleared.
  String? get() {
    if (_token == null || _expiresAt == null) return null;
    if (DateTime.now().isAfter(_expiresAt!)) {
      _token = null;
      _expiresAt = null;
      return null;
    }
    return _token;
  }

  /// Stores a new [token] in the cache with a fresh TTL.
  ///
  /// Replaces any previously cached token. The expiration is set to
  /// [ttl] from the current time.
  void set(String token) {
    _token = token;
    _expiresAt = DateTime.now().add(ttl);
  }

  /// Clears the cached token and expiration.
  ///
  /// After calling this, [get] returns `null` until [set] is called again.
  void clear() {
    _token = null;
    _expiresAt = null;
  }
}
